modules/project-factory/factory-projects-object.tf

```diff
@@ -95,107 +96,111 @@ locals {
       )
       parent                     = null
       prefix                     = null
-      service_encryption_key_ids = {}
+      service_encryption_key_ids = null
       storage_location           = null
-      tag_bindings               = {}
-      services                   = []
-      service_accounts           = {}
-      vpc_sc = merge({
-        perimeter_name    = null
-        perimeter_bridges = []
-        is_dry_run        = false
-        }, try(local._projects_config.data_overrides.vpc_sc, {
-          perimeter_name    = null
-          perimeter_bridges = []
-          is_dry_run        = false
-        })
-      )
-      logging_data_access = {}
+      tag_bindings               = null
+      services                   = null
+      service_accounts           = null
+      vpc_sc                     = try(local._projects_config.data_overrides.vpc_sc, null)
+      logging_data_access        = null
       },
       try(local._projects_config.data_overrides, {})
     )
   }
   _projects_output = {
-    for k, v in local._projects_input : lookup(v, "name", k) => merge(v, {
-      billing_account = try(coalesce(
+    # Semantics of the merges are:
+    # * if data_overrides.<field> is not null, use this value
+    # * if  _projects_inputs.<field> is not null, use this value
+    # * use data_default value, which if not set, will provide "empty" type
+    # This logic is easily implemented using coalesce, even on maps and list and allows to
+    # set data_overrides.<field> to "", [] or {} to ensure, that empty value is always passed, or do
+    # the same in _projects_input to prevent falling back to default value
+    for k, v in local._projects_input : k => merge(v, {
```

Comment:
I think we need to keep name as key. There are situations where the project data folder uses subfolders for organizing project files, and those are (rightly) stripped from they key. But then you risk duplicates in keys if you don't use the name as key.

Having the same file name across subfolders and then overriding with name inside the YAML is a valid pattern, and we should support it IMHO. The project factory is not a library, and the primary use case is supporting valid patterns in this module. If the net factory has an issue with this, it's a problem that needs to be addressed there.

Reply:
That's valid point, but risk of duplicates will happen before this stage, as here we already have a map on input. I'd opt to do that either while creating `local._projects_input` or post-processing `local._projects_output` if necessary and outside of this file.

Then we can provide guarantee, that keys here are stable between input and output, and you have the ability to cross-reference.

Reply:
I don't think that addresses support for the pattern described above. I don't see an issue using names as keys, control is still on the caller that defines both keys and names...